### ISSUE TITLE
seL4_ARM_ASIDControl_MakePool dosn't respect watermark

### ISSUE THREAD

**Issue by mullr** [seL4/seL4#128]
Most of the time when you use an untyped (with Untyped_Retype), you can pass a larger untyped from which some memory has already been taken, where the remaining capacity is tracked internally using watermark system described in the 'reusing memory' section of the manual. `seL4_ARM_ASIDControl_MakePool` appears to check the bit size of the untyped only, ignoring the watermark. 

Pseudocode repo: (the real one is in Rust, and currently closed-source)

```rust
fn repro(ut13: Untyped, asid_control: AsidControl) {
  let ut12a = ut13.retype(
     ut13,                          // _service
     api_object_seL4_UntypedObject, // type
     12,                            // size_bits
      ..., 
     1                              // num_objects
  );
  
  // Having consumed half of ut13, it's now effectively a ut12
  let ut12b = ut13;

  // this works, since we directly retyped ut12a to 12 bits
  assert(0 == seL4_ARM_ASIDControl_MakePool(
     asid_control,   // _service
     ut12a,          // untyped
      ...
   ));
  
  // this fails, even though ut12b has exactly 12 bits of memory left
  assert(0 == seL4_ARM_ASIDControl_MakePool(
     asid_control,   // _service
     ut12b,          // untyped
      ...
   ));
}
```

**Comment by kent-mcleod** [comment]
So you could still retype the remaining half of the 13-bit untyped into a 12-bit untyped and use that to create a second ASID pool.  The seL4_ARM_ASIDControl_MakePool invocation only accepts 12-bit untyped objects and doesn't attempt to provide all of the same features of an equivalent call to untype_retype.  Is your question more about whether seL4_ARM_ASIDControl_MakePool should support similar semantics to seL4_Untyped_Retype?

**Comment by mullr** [comment]
> Is your question more about whether seL4_ARM_ASIDControl_MakePool should support similar semantics to seL4_Untyped_Retype?

Yes. I can of course easily work around this in the manner you suggest. The issue is one of inconsistent API behavior, which led me to write a somewhat mysterious bug in my userland code. 

To give some context: I wrote a buddy allocator for Untypeds which built on the above behavior, as a trivial way to use half as many CNode slots. It actually only worked most of the time. I got "lucky" in all my initial tests and passed the correct side of the split to ASIDControl_MakePool. When the allocation order changed slightly, my luck ran out, and it took debugging into the kernel to find out what went wrong. 

For something as fundamental as Untyped objects, I would expect all specialization methods to follow the same allocation rules. (physical alignment / layout requirements notwithstanding, of course)

**Comment by pingerino** [comment]
@mullr I agree this is an inconsistency in the API.  Unfortunately fixing it requires changes to the seL4 proof, and our proof engineering team are overloaded at the moment and won't be able to get to this any time soon.

**Comment by mullr** [comment]
I see. Is there any guidance for submitting a patch for something like this? The C code itself should be quite easy, and presumably some portions of the proof must already exist in part for `Untyped_Retype`. I'd be happy to supply a patch for the source code, proofs, and Haskell model.

**Comment by pingerino** [comment]
Yes. As far as the C code goes you can upload a pull request here. Even better if you can also upload changes to see sel4test passing.

 Please see our contribution guidelines https://docs.sel4.systems/Contributing. 

I'll ping someone from the verification team to answer about contributing proofs/Haskell.

**Comment by lsf37** [comment]
Updating abstract spec and Haskell code would probably still be fairly straightforward, mostly mirroring the change in C. The same general contribution guidelines apply as for the C code, but for the l4v repository instead (https://github.com/seL4/l4v/).

If you don't have experience with Isabelle/HOL and interactive theorem proving, updating the proof for a kernel patch is probably too large for a first project, but we can probably take over from there and complete it. The main problem is finding someone who has enough time, because we're stretched a bit thin on project deadlines on the verification side at the moment.

I can't from the top of my head see any major issues with this particular proof update and API change, should be fine to do. It looks like we simply forgot to include it when we made the API change to incremental retype. It'd be a bit fiddly, and it'll be work, but I see no show-stoppers.

**Comment by jdub** [comment]
@lsf37 Could this be a simple-but-interesting enough change for a blog series about end-to-end seL4 development?

**Comment by lsf37** [comment]
@jdub Sorry, this seems to have dropped off my radar completely. Yes, it would be reasonably nice for such a blog post series. A lot of time has passed since then, would you still be interested in trying the verification angle?

**Comment by jdub** [comment]
@lsf37 I walked right into that, didn't I?

I have no idea how to even begin, but perhaps that's a useful perspective for an introductory text. I'll need a bit of a hand!

**Comment by lsf37** [comment]
:-) I'd be happy to help. For a change like this, it would make sense to start with the C code. After that, we'd have a look at the abstract and design specs, and then see how much of the proof breaks.

**Comment by kent-mcleod** [comment]
I think where this currently stands is:
1. It's not possible to use a UT greater than the ASID pool size to construct an ASID pool object in a single kernel invocation.
2.  It is possible with multiple invocations: First Retype parent UT into correctly sized UT, second construct pool with MakePool invocation, third delete intermediate UT cap.
3. There's a request to modify MakePool to support a larger UT cap and perform the internal allocation management from it.

Closing this until it ever becomes active again in the future.
